src/components/App/AppLoader.js

```diff
@@ -0,0 +1,128 @@
+import React, { useState, useEffect } from 'react'
+import PropTypes from 'prop-types'
+import LoadingScreen from '../LoadingScreen/LoadingScreen'
+
+const INTERNAL_APPS = ['home', 'permissions', 'apps', 'settings']
+
+const LOADING_ORG = 'Loading organization…'
+const LOADING_APPS = 'Loading apps…'
+const LOADING_APP = 'Loading {APP}…'
+const LOADING_READY = 'Ready.'
+
+// Pick the loading steps depending on what
+// is being loaded (internal vs.  external app)
+const LOADING_STEPS_INTERNAL = {
+  steps: [LOADING_ORG, LOADING_APPS, LOADING_READY],
+  stepIndex: ({ daoLoading, appsLoading }) =>
+    // pick the first truthy item as the current step
+    [daoLoading, appsLoading, true].findIndex(v => v),
+}
+
+const LOADING_STEPS_EXTERNAL = {
+  steps: [LOADING_ORG, LOADING_APPS, LOADING_APP, LOADING_READY],
+  stepIndex: ({ daoLoading, appsLoading, appLoading }) =>
+    // pick the first truthy item as the current step
+    [daoLoading, appsLoading, appLoading, true].findIndex(v => v),
+}
+
+function getLoadingSteps(instanceId) {
+  return INTERNAL_APPS.includes(instanceId)
+    ? LOADING_STEPS_INTERNAL
+    : LOADING_STEPS_EXTERNAL
+}
+
+function useLoadingStatus({
+  appLoading,
+  appsLoading,
+  currentAppName,
+  daoLoading,
+  instanceId,
+}) {
+  const [loadingSteps, setLoadingSteps] = useState(null)
+  const [loadingStepIndex, setLoadingStepIndex] = useState(0)
+
+  // Update the loading steps if `instanceId` changes
+  useEffect(() => {
+    const loadingSteps = getLoadingSteps(instanceId)
+    setLoadingSteps(loadingSteps)
+    setLoadingStepIndex(
+      loadingSteps.stepIndex({ daoLoading, appsLoading, appLoading })
+    )
+  }, [instanceId])
+
+  // Update the step index
+  useEffect(() => {
+    if (loadingSteps) {
```

Comment:
Should `loadingSteps` be a dependency here?

Reply:
Oh yes, thanks!